### DISCUSSION TITLE
[Feature request] Add relationLoaded() to Eloquent collection

### DISCUSSION THREAD

**Discussion by Vinze** [laravel/framework#49517] [General]
It would be nice if we can check if a relation is loaded on an Eloquent collection. See the following example:

```php
$projects = Project::with('client')->get();

$projects->relationLoaded('client'); // Method Illuminate\Database\Eloquent\Collection::relationLoaded does not exist.

// It works on a single Eloquent model
foreach ($projects as $project) {
    $project->relationLoaded('client'); // true
}

// Alternatively I could do it like this
$clientLoaded = $projects->count() && $projects->first()->relationLoaded('client'); // true
```

In my use case I'm rendering a table with projects which contains a 'client' column. This column should only be shown if the relation is actually loaded. For each project row I can check this with relationLoaded() but for the table header this isn't possible.

```html
<table>
  <thead>
    <tr>
      <th>Project</th>
      @if ($projects->relationLoaded('client')) <!-- This would be nice if it worked -->
        <th>Client</th>
      @endif
    </tr>
  </thead>
  <tbody>
    @foreach ($projects as $project)
      <tr>
        <td>{{ $project->name }}</td>
        @if ($project->relationLoaded('client'))
          <td>{{ $project->client->name }}</td>
        @endif
      </tr>
    @endforeach
  </tbody>
</table>
```

What do you guys think about this one?

**Comment by macropay-solutions** [comment]
From laravel 8 get on a relation:

    /**
     * Execute the query as a "select" statement.
     *
     * @param  array|string  $columns
     * @return \Illuminate\Database\Eloquent\Collection|static[]
     */
    public function get($columns = ['*'])
    {
        $builder = $this->applyScopes();

        // If we actually found models we will also eager load any relationships that
        // have been specified as needing to be eager loaded, which will solve the
        // n+1 query issue for the developers to avoid running a lot of queries.
        if (count($models = $builder->getModels($columns)) > 0) {
            $models = $builder->eagerLoadRelations($models);
        }

        return $builder->getModel()->newCollection($models);
    }

So after the get() all relations are eager loaded. For what do you need the loaded method?
if you want to see if it is null or model then:

        @if ($project->client instanceof Client)
          <td>{{ $project->client->name }}</td>
        @endif

**Reply by Vinze** [reply]
In my use case I have separated my application in a admin section and a client/customer section. So for example, I've built a 'projects' table as a view component and this gets included at multiple places. Now in the admin section I want to see all projects and to which client a project belongs. In the client section I also want to show the same table, but without the client column, as the client only sees his own projects. 

I can also solve this by passing a property to the component like this: `<x-projects.table :$projects section="admin" />` or run the `relationLoaded()` function on the first model of the collection. But it would be nice if it was possible to check on the collection level if a relation is loaded, not just on the model. That way I can simply check if a relation is loaded and show the columns/headings or else hide them.

**Reply by macropay-solutions** [reply]
Got it. Then 

    @if (property_exists($project, 'client'))
      <td>{{ $project->client?->name }}</td>
    @endif

or

    @if (isset($project->toArray()['client']))
      <td>{{ $project->client?->name }}</td>
    @endif

**Reply by Vinze** [reply]
Your example shows a single model, in which case I can use `$project->relationLoaded('client')`. Checking for null values isn't really a problem in this case. See the following:

```html
<table>
  <thead>
    <tr>
      <th>Project</th>
      @if ($projects->relationLoaded('client')) <!-- $projects collection -->
        <th>Client</th>
      @endif
    </tr>
  </thead>
  <tbody>
    ...
  </tbody>
</table>
```

In this example I'm only iterating over the $projects inside the tbody and here I can use relationLoaded. But in the thead I only need the table heading if a relation is loaded on a Eloquent collection. For this it would be nice if we could do `$projects->relationLoaded('client')` (mind the plural $projects)

**Reply by macropay-solutions** [reply]
Got it. Sorry, I missed the 's'. You can do that with a macro on the Collection until this gets included as new feature.

         Collection::macro(
            'relationLoaded',
            function (string $relation): bool {
                /** @var Collection $this */
                return ($model = $this->first()) instanceof Model && $model->relationLoaded($relation);
            }
        );

**Reply by Vinze** [reply]
Haven´t tried using a collection macro yet, will do that for now. Thanks!
